src/builder/QuarkBuilder.sol

```diff
@@ -266,53 +266,93 @@ contract QuarkBuilder {
         internal
         pure
     {
+        Actions.Action[] memory bridgeActions = findActionsOfType(actions, Actions.ACTION_TYPE_BRIDGE);
+        Actions.Action[] memory transferActions = findActionsOfType(actions, Actions.ACTION_TYPE_TRANSFER);
+
         uint256 plannedBridgeAmount = 0;
-        for (uint256 i = 0; i < actions.length; ++i) {
-            address tokenUsed;
-            uint256 amountUsed;
-            if (Strings.stringEqIgnoreCase(actions[i].actionType, Actions.ACTION_TYPE_TRANSFER)) {
-                Actions.TransferActionContext memory transferActionContext =
-                    abi.decode(actions[i].actionContext, (Actions.TransferActionContext));
-                tokenUsed = transferActionContext.token;
-                amountUsed = transferActionContext.amount;
-            } else if (Strings.stringEqIgnoreCase(actions[i].actionType, Actions.ACTION_TYPE_BRIDGE)) {
-                Actions.BridgeActionContext memory bridgeActionContext =
-                    abi.decode(actions[i].actionContext, (Actions.BridgeActionContext));
-                tokenUsed = bridgeActionContext.token;
-                amountUsed = bridgeActionContext.amount;
-                plannedBridgeAmount += amountUsed;
+        // Verify bridge actions is affordable, and update plannedBridgeAmount for verifying transfer actions
+        for (uint256 i = 0; i < bridgeActions.length; ++i) {
+            Actions.BridgeActionContext memory bridgeActionContext =
+                abi.decode(bridgeActions[i].actionContext, (Actions.BridgeActionContext));
+
+            if (bridgeActionContext.token == bridgeActions[i].paymentToken) {
+                // If the payment token is the transfer token and this is the target chain, we need to account for the transfer amount
+                uint256 paymentAssetBalanceOnChain = Accounts.sumBalances(
+                    Accounts.findAssetPositions(bridgeActionContext.token, bridgeActions[i].chainId, chainAccountsList)
+                );
+
+                // If its bridge step, check if user has enough balance to cover the bridge amount
+                if (paymentAssetBalanceOnChain < bridgeActions[i].paymentMaxCost + bridgeActionContext.amount) {
+                    revert MaxCostTooHigh();
+                }
             } else {
-                revert InvalidActionType();
+                // Just check payment token can cover the max cost
+                uint256 paymentAssetBalanceOnChain = Accounts.sumBalances(
+                    Accounts.findAssetPositions(
+                        bridgeActions[i].paymentToken, bridgeActions[i].chainId, chainAccountsList
+                    )
+                );
```

Comment:
This can be pulled out of the if statement and shared across the two cases.

Reply:
ok 👍